datafusion/core/src/datasource/listing/table.rs

```diff
@@ -1178,6 +1207,31 @@ impl ListingTable {
     }
 }
 
+/// Extension trait for FileSource to allow schema evolution support
+pub trait FileSourceExt {
```

Comment:
this is unfortunate (that we have something here that depends on parquet). Maybe we can add a `with_schema_adapter_factor` directly to `FileSource` 🤔